### ISSUE TITLE
Custom Layout - Grouping or Sorting

### ISSUE THREAD

**Issue by albertilagan** [F1bonacc1/process-compose#55]
Would be nice if we can define a custom group for each processes.
Or at the very least a `persistent` boolean flag and have those non persistent process be at the bottom of the list.

Sample 1:
```
Group 1
   - Process 1
   - Process 2
Group 2
   - Process 3
   - Process 4
```
ps: Group Name is not required a sample divider between groups is enough

Sample 2
```
Process 2
Process 3
Process 1 (non persistent)
Process 4 (non persistent)
```

Combination of both would be nice as well.

**Comment by faraonc** [comment]
This will be a nice addition. Sometimes I get confused when working with a monorepo.

**Comment by F1bonacc1** [comment]
Hi @albertilagan,

Thanks for the proposal.
In order to better understand your idea, can you please provide some of the use cases you have in mind?

This is what I have considered so far (feel free to correct any of those points):
1. Add an optional `group` flag to each process.
2. All the processes in each group are sorted lexicographically (or as they appear in the `config.yaml`?)
3. All the groups are sorted lexicographically.
4. All the processes without the new flag appear on top (or below?) of the grouped processes.

A few alternative approaches to consider:
1. Adding another `group` column and allowing to sort the table by any of the columns.
2. Showing the group name of the process in the process name `(group) name`, sorting everything lexicographically. Including processes without a group (will be at the bottom).
3. Adding another row to the table with the group name only.

**Comment by albertilagan** [comment]
> Hi @albertilagan,
> 
> Thanks for the proposal. In order to better understand your idea, can you please provide some of the use cases you have in mind?
> 
> This is what I have considered so far (feel free to correct any of those points):
> 
> 1. Add an optional `group` flag to each process.
> 2. All the processes in each group are sorted lexicographically (or as they appear in the `config.yaml`?)
> 3. All the groups are sorted lexicographically.
> 4. All the processes without the new flag appear on top (or below?) of the grouped processes.
> 
> A few alternative approaches to consider:
> 
> 1. Adding another `group` column and allowing to sort the table by any of the columns.
> 2. Showing the group name of the process in the process name `(group) name`, sorting everything lexicographically. Including processes without a group (will be at the bottom).
> 3. Adding another row to the table with the group name only.

This looks good actually. we actually don't use `process-compose` as is, we use it inside `devenv.nix` so I'm not too familiar with `config.yaml`

My take on this is
- sorting the process as they appear on the config would probably the easiest and solve most of our issue.
- The next addition is to introducing a `{ type: divider }`, together with `lexicographically`, if it see that `type=divider` we create empty rows (or however you wanna introduce space/divider).
- Lastly, adding `group` column as you suggested.

to answer this `In order to better understand your idea, can you please provide some of the use cases you have in mind?`.
our use case is we have a large monorepo and have a lot of commands/processes per app.

| PID | NAME | GROUP | STATUS |
|--------|--------|--------|--------|
| 123 | Caddy |  | Running |
| 456 | Redis |  | Running |
| 789 | Mailhog |  | Running |
| 1011 | Postgres |  | Running |
|--------|--------|--------|--------|
| 1213 | Proc1 | Group1 | Running |
| 1415 | Proc2 | Group1 | Disabled | 
|--------|--------|--------|--------|
| 1617 | Proc3 | Group2 | Running |
| 1819 | Proc4 | Group2 | Disabled | 

or

| PID | NAME | STATUS |
|--------|--------|--------|
| 123 | Caddy | Running |
| 456 | Redis | Running |
| 789 | Mailhog | Running |
| 1011 | Postgres | Running |
|--------|--------|--------|
| 1213 | Proc1 | Running |
| 1415 | Proc2 | Disabled | 
|--------|--------|--------|
| 1617 | Proc3 | Running |
| 1819 | Proc4 | Disabled |

**Comment by albertilagan** [comment]
@F1bonacc1  just wanna clarify that if you want to introduce `group` field and have spacing automatically applied in between groups we don't need `{ type: divider }` anymore.

so either `{ type: divider }` or `{ group: <value> }`, whatever you find more suitable.

**Comment by F1bonacc1** [comment]
@albertilagan, @faraonc 
In [v0.51.0](https://github.com/F1bonacc1/process-compose/releases/tag/v0.51.0) I added a concept of namespaces.
For now, they are only used for sorting
If there will be enough interest, I will also add the ability to filter by namespace.
If not defined the namespace will be `default`.

**Comment by albertilagan** [comment]
> @albertilagan, @faraonc In [v0.51.0](https://github.com/F1bonacc1/process-compose/releases/tag/v0.51.0) I added a concept of namespaces. For now, they are only used for sorting If there will be enough interest, I will also add the ability to filter by namespace. If not defined the namespace will be `default`.

Do we have an example on this feature yet?
Also I'm using `devenv` will it work if I do `devenv update`? or do we need `devenv` to update their `process-compose` reference?

**Comment by F1bonacc1** [comment]
The vendor hash changes only when my dependencies change. They didn't change from v0.45.0 to v0.51.0.

**Comment by thenonameguy** [comment]
> > @albertilagan, @faraonc In [v0.51.0](https://github.com/F1bonacc1/process-compose/releases/tag/v0.51.0) I added a concept of namespaces. For now, they are only used for sorting If there will be enough interest, I will also add the ability to filter by namespace. If not defined the namespace will be `default`.
> 
> 
> 
> ~~Do we have an example on this feature yet?~~
> 
> ~~Also I'm using `devenv` will it work if I do `devenv update`? or do we need `devenv` to update their `process-compose` reference?~~
> 
> 
> 
> How can I contribute to this? I can't figure out how to update the `vendorHash`. [Process Compose v0.45.0](https://github.com/NixOS/nixpkgs/commit/f2a43c25554c5e5fa6b61c09d3f0e08398c03b3a)

I maintain the nixpkgs package of process-compose. A bot will usually pick the new release up and create a PR. It will land in a few days in nixpkgs master. Then you just have to update your nixpkgs lock in devenv (via devenv.yaml).

Until then you can add the flake of this github repo and use it directly (read: compile it yourself).

**Comment by F1bonacc1** [comment]
Hi @albertilagan,

It seems that v0.51.0 is now in master:
https://github.com/NixOS/nixpkgs/blob/e9e3f2e7362aba3ef709173d1777f0b893fc0ebf/pkgs/applications/misc/process-compose/default.nix#L11

Can you please confirm that it works for you as expected?

**Comment by albertilagan** [comment]
> Hi @albertilagan,
> 
> 
> 
> It seems that v0.51.0 is now in master:
> 
> https://github.com/NixOS/nixpkgs/blob/e9e3f2e7362aba3ef709173d1777f0b893fc0ebf/pkgs/applications/misc/process-compose/default.nix#L11
> 
> 
> 
> Can you please confirm that it works for you as expected?

Yup was able to test it. Works as expected even though there's no spacing between namespaces, but I guess it's restricted by the table you're using. 

That being said it would be nice if the default sort applies to namespace or atleast being to define default sort column in config.

**Comment by thenonameguy** [comment]
> > > > @albertilagan, @faraonc In [v0.51.0](https://github.com/F1bonacc1/process-compose/releases/tag/v0.51.0) I added a concept of namespaces. For now, they are only used for sorting If there will be enough interest, I will also add the ability to filter by namespace. If not defined the namespace will be `default`.
> > > 
> > > 
> > > ~Do we have an example on this feature yet?~
> > > ~Also I'm using `devenv` will it work if I do `devenv update`? or do we need `devenv` to update their `process-compose` reference?~
> > > How can I contribute to this? I can't figure out how to update the `vendorHash`. [Process Compose v0.45.0](https://github.com/NixOS/nixpkgs/commit/f2a43c25554c5e5fa6b61c09d3f0e08398c03b3a)
> > 
> > 
> > I maintain the nixpkgs package of process-compose. A bot will usually pick the new release up and create a PR. It will land in a few days in nixpkgs master. Then you just have to update your nixpkgs lock in devenv (via devenv.yaml).
> > Until then you can add the flake of this github repo and use it directly (read: compile it yourself).
> 
> @thenonameguy do we have update on this? :)

https://github.com/NixOS/nixpkgs/commit/691a2ad003722cbf965a8f19f93302f13e6b0401 
Better late than never :)

**Comment by F1bonacc1** [comment]
> > Hi @albertilagan,
> > It seems that v0.51.0 is now in master:
> > https://github.com/NixOS/nixpkgs/blob/e9e3f2e7362aba3ef709173d1777f0b893fc0ebf/pkgs/applications/misc/process-compose/default.nix#L11
> > Can you please confirm that it works for you as expected?
> 
> Yup was able to test it. Works as expected even though there's no spacing between namespaces, but I guess it's restricted by the table you're using.
> 
> That being said it would be nice if the default sort applies to namespace or atleast being to define default sort column in config.

@albertilagan 
You can now create an alias that will use namespace (or any other column name) as the default sort column:
```shell
alias pc='process-compose -S namespace'
pc
```
Available from version v0.69.0.
